tools/symbols_map_generator.py

```diff
@@ -0,0 +1,343 @@
+                    for child in derived.get_children():
+                        if child.kind != clang.cindex.CursorKind.CXX_BASE_SPECIFIER:
+                            continue
+
+                        class_or_struct_node = clang.cindex.conf.lib.clang_getCursorDefinition(child)
+                        if class_or_struct_node is None:
+                            continue
+
+                        base_classes.append(class_or_struct_node)
+                        if not clang.cindex.conf.lib.clang_isVirtualBase(child):
+                            continue
+
+                        # Search the immediate base classes for the function name
+                        for other_child in class_or_struct_node.get_children():
+                            if (other_child.displayname == node.displayname):
+                                result.add(f"virtual?thunk?to?{namespace_str};")
+                                return True
+
+                    # Looks like it wasn't in any of our direct ancestors, so let's 
+                    # try their ancestors too
+                    for base in base_classes:
+                        if search_class_hierarchy_for_virtual_thunk(base):
+                            return True
+                        
+                    return False
+                    
+                search_class_hierarchy_for_virtual_thunk(node.semantic_parent)
+                            
+
+    # Traverse down the tree if we can
+    if (node.kind == clang.cindex.CursorKind.TRANSLATION_UNIT
+        or node.kind == clang.cindex.CursorKind.CLASS_DECL
+        or node.kind == clang.cindex.CursorKind.STRUCT_DECL
+        or node.kind == clang.cindex.CursorKind.NAMESPACE):
+        for child in node.get_children():
+            traverse_ast(child, filename, result)
+
+    return result
+
+
+def process_directory(directory: str, search_dirs: Optional[list[str]]) -> list[str]:
+    current_set = set()
+
+    files_dir = get_absolute_path_from_project_path(directory)
+    _logger.debug(f"Processing external directory: {files_dir}")
+    files = files_dir.rglob('*.h')
+    search_variables = []
+    for dir in search_dirs:
+        search_variables.append(f"-I{get_absolute_path_from_project_path(dir).as_posix()}")
+    args = ['-std=c++20', '-x', 'c++-header', '-nostdlibinc'] + search_variables
+    for file in files:
+        idx = clang.cindex.Index.create()
+        tu = idx.parse(
+            file.as_posix(), 
+            args=args,  
+            options=0)
+        root = tu.cursor
+        list(traverse_ast(root, file.as_posix(), current_set))
+
+    result = list(current_set)
+    result.sort()
+    return result 
+
+
+def output_symbols_to_file(symbols: list[str], library: str, version: str, output_file: str):
+    joint_symbols = "\n    ".join(symbols)
+    output_str = f'''{library.upper()}_{version} {"{"}
+global:
+  extern "C++" {"{"}
+    {joint_symbols}
+  {'};'}
+local: *;
+{'}'};'''
```

Comment:
This looks like it is outputting a single stanza, which is wrong in the general case. Except when ABI breaks symbols should be in a stanza that matches the minor release in which they were introduced. (Have a look at `src/common/symbols.map` for an example.)

That means we need to reproduce "old" stanzas verbatum and strip any symbols those provide from the outputt to the "current" stanza. (See `src/miral/regenerate-miral-symbols-map.py:566` for the corresponding logic)

Reply:
> That means we need to reproduce "old" stanzas verbatum and strip any symbols those provide from the outputt to the "current" stanza. (See src/miral/regenerate-miral-symbols-map.py:566 for the corresponding logic)

An (more invasive) alternative we *could* choose here would be to tag everything with `[[mir:since("$VERSION")]]`, pull that out in the generator, and generate the set of stanzas based on the since-version.

That's probably more work than we want to do now, though.

Reply:
> That's probably more work than we want to do now, though.

It also adds the complexity of needing to group any pre-ABI break symbols together.

Reply:
Removed that :)